// nodemondotenv/server/controllers/orderController.js
const { Orders, Users } = require('../models/models');

class OrderController {
    async getAllOrders(req, res) {
        try {
            const orders = await Orders.findAll({
                include: [{ model: Users, attributes: ['name', 'surname', 'patronymic'] }],
                order: [['createdAt', 'DESC']]
            });
            return res.json(orders);
        } catch (error) {
            console.error('Ошибка при получении заказов:', error);
            return res.status(500).json({ message: 'Ошибка при получении заказов' });
        }
    }

    async updateOrderStatus(req, res) {
        const { orderId } = req.params;
        const { status, reason } = req.body;

        try {
            const order = await Orders.findOne({ where: { id: orderId } });
            if (!order) {
                return res.status(404).json({ message: 'Заказ не найден' });
            }

            order.status = status;
            if (reason) {
                order.reason = reason;
            }

            await order.save();
            return res.json({ message: 'Статус заказа обновлен' });
        } catch (error) {
            console.error('Ошибка при обновлении статуса заказа:', error);
            return res.status(500).json({ message: 'Ошибка при обновлении статуса заказа' });
        }
    }

    async deleteOrder(req, res) {
        const { orderId } = req.params;

        try {
            const order = await Orders.findOne({ where: { id: orderId } });
            if (!order) {
                return res.status(404).json({ message: 'Заказ не найден' });
            }

            await order.destroy();
            return res.json({ message: 'Заказ удален' });
        } catch (error) {
            console.error('Ошибка при удалении заказа:', error);
            return res.status(500).json({ message: 'Ошибка при удалении заказа' });
        }
    }
}

module.exports = new OrderController();
